pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -171,6 +211,140 @@ func referencesDiffs(old []metatypes.OwnerReference, new []metatypes.OwnerRefere
 	return added, removed
 }
 
+func shouldOrphanDependents(e event, accessor meta.Object) bool {
+	// The delta_fifo may combine the creation and update of the object into one
+	// event, so we need to check AddEvent as well.
+	if e.oldObj == nil {
+		if accessor.GetDeletionTimestamp() == nil {
+			return false
+		}
+	} else {
+		oldAccessor, err := meta.Accessor(e.oldObj)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("cannot access oldObj: %v", err))
+			return false
+		}
+		// ignore the event if it's not updating DeletionTimestamp from non-nil to nil.
+		if accessor.GetDeletionTimestamp() == nil || oldAccessor.GetDeletionTimestamp() != nil {
+			return false
+		}
+	}
+	return sets.NewString(accessor.GetFinalizers()...).Has(api.OrphanFinalizer)
+}
+
+func (gc *GarbageCollector) orhpanDependents(owner *node, dependents []*node, retries int) error {
+	var failedDependents []objectReference
+	for _, dependent := range dependents {
+		deleteOwnerRefPatch := fmt.Sprintf(`{"metadata":{"ownerReferences":[{"$patch":"delete","uid":"%s"}]}}`, owner.identity.UID)
+		patched := false
+		for count := 0; count < retries; count++ {
+			_, err := gc.patchObject(dependent.identity, []byte(deleteOwnerRefPatch))
+			// note that if the target ownerReference doesn't exist in the
+			// dependent, strategic merge patch will NOT return an error.
+			if err == nil {
+				patched = true
+				break
+			} else {
+				utilruntime.HandleError(fmt.Errorf("orphaning %s failed with %v", dependent.identity, err))
+			}
+		}
+		if !patched {
+			failedDependents = append(failedDependents, dependent.identity)
+		}
+	}
+	if len(failedDependents) != 0 {
+		return fmt.Errorf("failed to orphan dependents %v, please manually orphan them and finalize the owner %s", failedDependents, owner.identity)
+	}
+	return nil
+}
+
+// TODO: Using Patch when strategicmerge supports deleting an entry from a
+// slice of a base type.
+func (gc *GarbageCollector) removeOrphanFinalizer(owner *node, retries int) error {
+	updated := false
+	for count := 0; count < retries; count++ {
```

Comment:
A update operation can easily get false conflicts, so I let it retry immediately if the update fails with a conflict. If it's other kinds of error, I'll return error immediately and rely on the outer queue/retry thingy.